### PR TITLE
[Sprint 52] 1.2.x XD-3125, XD-3126 refactor and enhance serialization configuration

### PR DESCRIPTION
**Pull request by dturanski**
This PR, motivated by XD-3125 and XD-3126, refactors and enhances Kryo serialization. Here is a summary of the changes

* The "codec" bean, CompositeCodec {TupleCodec, FileCodec, PojoCodec} is replaced by PojoCodec with custom serializers configured for Tuple and File.
* PojoCodec is wired with List<AbstractKryoRegistrar>. Previously, KryoRegistrar was a singleton that was configured only if a user wished to register domain objects with a unique ID.  Now, PojoCodec may be configured with multiple `AbstractKryoRegistrar`s which are merged and validated, (checked for conflicting registrations) at start up. Kryo Registrations are logged at the INFO level and any conflicts will result in a SerializationExeption at start up. XD provides `TupleKryoRegistrar` and `FileKryoRegistrar` and users may provide additional ones and register them in the SharedServerContext. Spring will "autowire" any `AbstractKryoRegistrar`s in a known base package or XML location.  Note that annotation based autowiring was tried but turned out to be less clean in this case.
* codec.xml replaced with Java config in consideration of support for potentially implementing an alternate "codec" without forcing xml config. Related to https://jira.spring.io/browse/SPR-7341. Currently "codec" is configured in `SharedServerContextConfiguration` and `@ConditionalOnMissingBean(name = "codec")`  Currently the message bus configuration requires a bean named "codec". That could be improved via autowiring as well, but looks like the DI will require additional refactoring in the class hierarchy. 
* Added `RegistrationKryoRegistrar` to support user defined Kryo Serializers. This takes a `List<Registration>`, where Registration is a native Kryo type associating (type, Serializer, ID). 
* The existing extension point for serializers is META-INF/spring-xd/bus/ext/*.xml. This is retained for backward compatibility. Component scanning is also enabled for "spring.xd.bus.ext". This package is hard coded currently in SharedServerContextConfiguration. It didn't seem worth making the base package configurable for this as we did for PluginContext extensions (it seems that feature has been used very little, if at all).  I would consider a different package name, but currently it mirrors, more or less, the xml path.  

So the idea is that in order to use one or more custom serializers,  a user would need to install a jar file with their POJO(s), corresponding Serializer(s), and an @Configuration class located in spring.xd.bus.ext to configure a RegistrationKryoRegistrar, as is done in  `ExtensionsTest` and spring-xd-dirt/src/test/java/spring.xd.bus.ext/CustomKryoRegistrarConfig. 

Note that a couple of alternatives are supported by Kryo already. One is to annotate the POJO class with @DefaultSerializer. Registering the class with a unique ID is also recommended for optimal performance. This also requires installing a custom jar to XD, but requires access to the POJO source. Additionally, users may have the POJO implement (Kryo) Serializable. This can result in 50% improvement but has been observed to be about 10% slower than registering a custom serializer.

### CODE REVIEW

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/plugins/spark/streaming/MessageBusConfiguration.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
 @Configuration
 @Import(PropertyPlaceholderAutoConfiguration.class)
 @ImportResource({"classpath*:" + ConfigLocations.XD_CONFIG_ROOT + "bus/${XD_TRANSPORT}-bus.xml",
-		ConfigLocations.XD_CONFIG_ROOT + "bus/codec.xml"})
+		ConfigLocations.XD_CONFIG_ROOT + "bus/codec-hide.xml"})
```
Comment (dturanski) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/plugins/spark/streaming/MessageBusConfiguration.java]: whoops!

**Comment by dturanski** [comment]
Just realized there's an issue with the spark MessageBusConfig. Will take some work to get around it.

File: spring-xd-integration-test/src/main/java/org/springframework/xd/integration/util/StreamUtils.java [hunk 2]
```diff
@@ -499,6 +499,26 @@ public static boolean waitForMetric(String name, URL adminServer, int waitTime)
 		return extractPidsFromJPS(response.getOutput());
 	}
```
Comment (markpollack) [inline spring-xd-integration-test/src/main/java/org/springframework/xd/integration/util/StreamUtils.java]: This looks like it was pulled in with some other unrelated changes under integration-test project

File: spring-xd-integration-test/src/test/java/org/springframework/xd/integration/test/AbstractIntegrationTest.java [hunk 3]
```diff
@@ -265,28 +265,34 @@ public void setupSourceDataFiles(String sourceDir, String fileName, String data)
...
-
+		boolean result = false;
 		if (xdEnvironment.isOnEc2()) {
-			StreamUtils.createDataFileOnRemote(xdEnvironment.getPrivateKey(), host, sourceDir, fileName, data,
+			result = StreamUtils.createDataFileOnRemote(xdEnvironment.getPrivateKey(), host, sourceDir, fileName, data,
 					WAIT_TIME);
 		}
 		else {
 			try {
+				File dirs = new File(sourceDir);
```
Comment (markpollack) [inline spring-xd-integration-test/src/test/java/org/springframework/xd/integration/test/AbstractIntegrationTest.java]: This looks like it was pulled in with some other unrelated changes under integration-test project

File: spring-xd-integration-test/src/test/java/org/springframework/xd/integration/test/FileJdbcTest.java [hunk 4]
```diff
@@ -37,6 +39,8 @@
 public class FileJdbcTest extends AbstractJobTest {
 
 	private final static String DEFAULT_FILE_NAME = "filejdbctest";
+	private final static String DEFAULT_DIRECTORY = XdEnvironment.RESULT_LOCATION +
```
Comment (markpollack) [inline spring-xd-integration-test/src/test/java/org/springframework/xd/integration/test/FileJdbcTest.java]: This looks like it was pulled in with some other unrelated changes under integration-test project

File: spring-xd-yarn/site/config/servers.yml [hunk 5]
```diff
@@ -58,7 +58,7 @@ spring:
 #      mapreduce.application.framework.path: '/phd/apps/3.0.0.0-249/mapreduce/mapreduce.tar.gz#mr-framework'
 #      mapreduce.application.classpath: '$PWD/mr-framework/hadoop/share/hadoop/mapreduce/*:$PWD/mr-framework/hadoop/share/hadoop/mapreduce/lib/*:$PWD/mr-framework/hadoop/share/hadoop/common/*:$PWD/mr-framework/hadoop/share/hadoop/common/lib/*:$PWD/mr-framework/hadoop/share/hadoop/yarn/*:$PWD/mr-framework/hadoop/share/hadoop/yarn/lib/*:$PWD/mr-framework/hadoop/share/hadoop/hdfs/*:$PWD/mr-framework/hadoop/share/hadoop/hdfs/lib/*:/usr/phd/3.0.0.0-249/hadoop/lib/hadoop-lzo-0.6.0.3.0.0.0-249.jar:/etc/hadoop/conf/secure'
 
-# Hortonworks HDP 2.2 (hdp22):
+# Hortonworks HDP 2.2.4.2 (hdp22):
```
Comment (markpollack) [inline spring-xd-yarn/site/config/servers.yml]: Looks like an unrelated commit

File: spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/CompositeKryoRegistrar.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * A {@link org.springframework.xd.dirt.integration.bus.serializer.kryo.KryoRegistrar} that delegates and validates
+ * registrations across all components.
+ * @author David Turanski
+ * @since 1.2
+ */
+public class CompositeKryoRegistrar extends AbstractKryoRegistrar {
+
+	private final List<AbstractKryoRegistrar> delegates;
```
Comment (markpollack) [inline spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/CompositeKryoRegistrar.java]: See if can refactor to use a modified KryoRegistrar interface vs. abstract base class.

**Comment by dturanski** [comment]
Close this and create a new PR to eliminate extraneous commits
